graphein/datasets/pdb_data.py

```diff
@@ -0,0 +1,1092 @@
+            )
+        log.info("Done extracting chains")
+
+        # Flatten list of paths
+        return [Path(num) for sublist in paths for num in sublist]
+
+    def reset(self) -> pd.DataFrame:
+        """Reset the dataset to the original DataFrame source.
+
+        :return: The source dataset DataFrame.
+        :rtype: pd.DataFrame
+        """
+        self.df = self.source.copy()
+        return self.df
+
+    def split_df_proportionally(
+        self,
+        df: pd.DataFrame,
+        splits: List[str],
+        split_ratios: List[float],
+        assign_leftover_rows_to_split_n: int = 0,
+        random_state: int = 42,
+    ) -> Dict[str, pd.DataFrame]:
+        """Split the provided DataFrame iteratively according to given proportions.
+
+        :param df: DataFrame to split.
+        :type df: pd.DataFrame
+        :param splits: Names of splits into which to divide the provided DataFrame.
+        :type splits: List[str]
+        :param split_ratios: Ratios by which to split the provided DataFrame.
+        :type split_ratios: List[float]
+        :param assign_leftover_rows_to_split_n: To which split to assign leftover rows,
+            defaults to ``0``.
+        :type assign_leftover_rows_to_split_n: int, optional
+        :param random_state: Random seed to use for DataFrame splitting, defaults to
+            ``42``.
+        :type random_state: int, optional
+
+        :return: Dictionary of DataFrame splits.
+        :rtype: Dict[str, pd.DataFrame]
+        """
+        # Calculate the size of each split
+        split_sizes = [int(len(df) * ratio) for ratio in split_ratios]
+
+        # Assign leftover rows to a specified split
+        num_remaining_rows = len(df) - sum(split_sizes)
+        if num_remaining_rows > 0:
+            split_sizes[assign_leftover_rows_to_split_n] += num_remaining_rows
+
+        # Without replacement, randomly shuffle rows within the input DataFrame
+        df_sampled = df.sample(
+            frac=1.0, replace=False, random_state=random_state
+        )
+
+        # Split DataFrames
+        start_idx = 0
+        df_splits = {}
+        for split_index, split_size in enumerate(split_sizes):
+            split = splits[split_index]
+            end_idx = start_idx + split_size
+            df_split = df_sampled.iloc[start_idx:end_idx]
+            df_splits[split] = df_split
+            start_idx = end_idx
+
+        # Ensure there are no duplicated rows between splits
+        all_rows = pd.concat([df_splits[split] for split in splits])
+        assert len(all_rows) == len(
+            df
+        ), "Number of rows changed during split operations."
+        assert len(
+            all_rows.drop(self.list_columns, axis=1).drop_duplicates()
+        ) == len(df), "Duplicate rows found in splits."
+
+        return df_splits
+
+    def merge_df_splits(
+        self,
+        first_df_split: pd.DataFrame,
+        second_df_split: pd.DataFrame,
+        df_primary_key: str = "id",
```

Comment:
Do we want to keep splits as attributes & provide an update arg here?

Reply:
I think this relates to your comment on the point I made in the initial pull request post above, in that this would allow us to more easily manage to which split a given sequence (i.e., chain) belongs at any given point.

Reply:
Towards this end, in [this commit](https://github.com/a-r-j/graphein/pull/272/commits/1878f78e541266b02d736e51799e009021871563), I have added preliminary support for tracking to which "split" a given sequence (i.e., PDB chain) belongs.

Reply:
I like this! So I think this can open some additional functionality for split-specific operations.

E.g:

`manager.to_fasta(split="train")`

Or `manager.longer_than(100, split="val")`